ledger/src/blockstore_meta.rs

```diff
@@ -262,6 +299,226 @@ impl ShredIndex {
         self.index.insert(index);
     }
+
+    #[cfg(test)]
+    fn remove(&mut self, index: u64) {
+        self.index.remove(&index);
+    }
+}
+
+/// A bitvec (`Vec<u8>`) of shred indices, where each u8 represents 8 shred indices.
+///
+/// The current implementation of [`ShredIndex`] utilizes a [`BTreeSet`] to store
+/// shred indices. While [`BTreeSet`] remains efficient as operations are amortized
+/// over time, the overhead of the B-tree structure becomes significant when frequently
+/// serialized and deserialized. In particular:
+/// - **Tree Traversal**: Serialization requires walking the non-contiguous tree structure.
+/// - **Reconstruction**: Deserialization involves rebuilding the tree in bulk,
+///   including dynamic memory allocations and re-balancing nodes.
+///
+/// In contrast, our bit vec implementation provides:
+/// - **Contiguous Memory**: All bits are stored in a contiguous array of u64 words,
+///   allowing direct indexing and efficient memory access patterns.
+/// - **Direct Range Access**: Can load only the specific words that overlap with a
+///   requested range, avoiding unnecessary traversal.
+/// - **Simplified Serialization**: The contiguous memory layout allows for efficient
+///   serialization/deserialization without tree reconstruction.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct ShredIndexV2 {
+    #[serde(with = "serde_bytes")]
+    index: Vec<u8>,
+    num_shreds: usize,
+}
+
+impl Default for ShredIndexV2 {
+    fn default() -> Self {
+        Self {
+            index: vec![0; Self::MAX_WORDS_PER_SHRED],
+            num_shreds: 0,
+        }
+    }
+}
+
+type ShredIndexV2Word = u8;
+impl ShredIndexV2 {
+    const SIZE_OF_WORD: usize = std::mem::size_of::<ShredIndexV2Word>();
+    const WORD_BITS: usize = Self::SIZE_OF_WORD * 8;
+    const MAX_WORDS_PER_SHRED: usize = MAX_DATA_SHREDS_PER_SLOT.div_ceil(Self::WORD_BITS);
+
+    pub fn num_shreds(&self) -> usize {
+        self.num_shreds
+    }
+
+    fn index_and_mask(index: u64) -> (usize, ShredIndexV2Word) {
+        let word_idx = index as usize / Self::WORD_BITS;
+        let bit_idx = index as usize % Self::WORD_BITS;
+        let mask = 1 << bit_idx;
+        (word_idx, mask as ShredIndexV2Word)
+    }
+
+    #[cfg(test)]
+    fn remove(&mut self, index: u64) {
+        assert!(
+            index < MAX_DATA_SHREDS_PER_SLOT as u64,
+            "index out of bounds. {index} >= {MAX_DATA_SHREDS_PER_SLOT}"
+        );
+
+        let (word_idx, mask) = Self::index_and_mask(index);
+
+        if self.index[word_idx] & mask != 0 {
+            self.index[word_idx] ^= mask;
+            self.num_shreds -= 1;
+        }
+    }
+
+    #[allow(unused)]
+    pub(crate) fn contains(&self, idx: u64) -> bool {
+        if idx >= MAX_DATA_SHREDS_PER_SLOT as u64 {
+            return false;
+        }
+        let (word_idx, mask) = Self::index_and_mask(idx);
+        (self.index[word_idx] & mask) != 0
+    }
```

Comment:
We have some similar code in other functions, was thinking about how we might consolidate but I think it is clunky and there is so little "duplicate" code that what you have is good as-is